### PR TITLE
Disallow interactions by blocked users; some interaction API refactoring (fixes #7971) {Continued in #8755}

### LINKED ISSUE #7971: flags.chatRevoked does not prevent user sending private messages (PMs)
When a user's chat permissions need to be revoked, they are given a true value for `flags.chatRevoked`. This stops them posting to any guild or the Tavern.

It should also stop them sending private messages to any player.

### PR DESCRIPTION
**Pull request by CrystallineCat**
Fixes #7971

This PR changes the way objections to user-to-user interactions are handled (e.g. chat revocation or blocks). It includes some refactoring to the API to make it more flexible and allows the UI to query the estimated success of interactions. No UI changes are done at this time.

API refactoring:
- Factors out the checks in `api.sendPrivateMessage` and `api.transferGems` to see if the interaction is allowed.
- The checks are now in `User.getObjectionsToInteractionIfAny`; the method is designed to be easily extensible to new checks or new interactions.
- To allow the UI to check whether an interaction is allowed without actually preforming it, there is a new function `api.getObjectionsToInteractionIfAny`; it only applies interactions initiated by the logged in user and can not be used to query interactions by other users.
- All objections to an interaction now use NotAuthorized instead of NotFound.
- All objection strings are now translatable.

Additional checks:
- Both `api.sendPrivateMessage` and `api.transferGems` now are now blocked by `sender.flags.chatRevoked`.
- `api.transferGems` is now blocked by either user blocking the other.

Tests:
- Tests for the additional checks and the new API have been added.

---

UUID: fea892e5-ac99-4b03-83ac-d01a2154e2af

### CODE REVIEW

File: test/api/v3/integration/members/GET-objections_to_interaction_if_any.test.js
Comment (khipkin) [inline test/api/v3/integration/members/GET-objections_to_interaction_if_any.test.js]: Suggestion: delete this line. :)

File: website/server/models/user/methods.js
Comment (khipkin) [inline website/server/models/user/methods.js]: Suggestion: add this string to a file in the common/locales/en directory with an "interaction" parameter.  The known interactions can be manually added (probably) because there are a known, finite number of allowed interactions.

(You should verify this with someone on the Habitica staff before implementing though. :smile: )

File: test/api/v3/integration/members/POST-transfer_gems.test.js
Comment (khipkin) [inline test/api/v3/integration/members/POST-transfer_gems.test.js]: Maybe it's just me, but this variable name is confusing me.

File: test/api/v3/integration/members/POST-transfer_gems.test.js
Comment (khipkin) [inline test/api/v3/integration/members/POST-transfer_gems.test.js:90]: What about if someone sends gems to someone with chat privileges revoked?

File: website/server/controllers/api-v3/members.js
Comment (khipkin) [inline website/server/controllers/api-v3/members.js]: Maybe allow the callee to throw this message ifneedbe? And don't catch it here (in the caller)?

File: website/server/models/user/methods.js
Comment (khipkin) [inline website/server/models/user/methods.js]: Technically, if the given interaction was attempted, only one error would be thrown, since it would be pointless to keep trying once a failure condition was found.  Is it really necessary to return ALL issues with an interaction?  Why not just the first one found (if a problem exists) or none (if the interaction is allowed)?

File: website/server/models/user/methods.js
Comment (khipkin) [inline website/server/models/user/methods.js]: What does NB stand for?

Comment (Alys) [reply website/server/models/user/methods.js]: Nota Bene ("note well"). Let's remove "NB" since it's not necessary and few of us are ancient Romans. :) Sorry, CrystallineCat - entirely my fault. I hadn't thought that it wasn't widely known.

File: test/api/v3/integration/members/POST-transfer_gems.test.js
Comment (Alys) [reply test/api/v3/integration/members/POST-transfer_gems.test.js:90]: If the recipient has chat permissions revoked, sending gems to them is allowed.

If the sender has site-wide chat privileges revoked, or has been personally blocked by the recipient, then they should not be able to send gems.

Does that cover what you're asking and/or what the code is doing / not doing?  (I haven't checked the code in detail, sorry. :( )

Comment (khipkin) [reply test/api/v3/integration/members/POST-transfer_gems.test.js:90]: I was just suggesting that test coverage for the above situation be added. :smile_cat:

File: website/server/models/user/methods.js
Comment (crookedneighbor) [inline website/server/models/user/methods.js]: Why disable eslint here?

File: website/server/models/user/methods.js
Comment (crookedneighbor) [inline website/server/models/user/methods.js]: Can we remove the `IfAny` from the name? I don't think that's necessary.

File: website/server/models/user/methods.js
Comment (crookedneighbor) [inline website/server/models/user/methods.js]: Lets avoid one character variables. `filter(check => check !== 'aways');`  should do it.

File: website/server/models/user/methods.js
Comment (crookedneighbor) [reply website/server/models/user/methods.js]: This would be a developer error, not something that should be exposed to the user, so it shouldn't be added to the locale files.

Comment (khipkin) [reply website/server/models/user/methods.js]: In that case, how should it be handled?

Comment (crookedneighbor) [reply website/server/models/user/methods.js]: Just like this. Throwing a normal error.

File: website/server/models/user/methods.js
Comment (CrystallineCat) [reply website/server/models/user/methods.js]: Actually, it's not your fault, the NB wasn't in your original line. I added it because the comment looked a bit out of place to me without any code to follow it. I didn't consider it might confuse people, so I agree it's better to remove it. :)

File: website/server/models/user/methods.js
Comment (CrystallineCat) [reply website/server/models/user/methods.js]: The checks are all of the form `(sndr, rcvr) => ...`, due to being iterated over and called with the same signature, but not all of them use both parameters. This leads to spurious `no-unused-vars` issues.

I think following the linter here would be wrong and make the code worse, not better. I've added a comment stating the reason for the disable.

File: website/server/models/user/methods.js
Comment (Alys) [reply website/server/models/user/methods.js]: Oh! Huh. :) "NB" is something I use myself so often I just assumed I'd specified that it should be used. :)

File: website/server/models/user/methods.js
Comment (CrystallineCat) [reply website/server/models/user/methods.js]: I've thought about that, but I'm not sure it's the way to go. I'd rather change the error reporting code to be able to report multiple issues -- the new API function already does that.

Reporting only one when there are multiple seems to me to give a skewed view of the situation, kinda like "fix this one error and it will work", while reporting all gives the full view -- but maybe I'm reading too much into things here.

File: website/server/controllers/api-v3/members.js
Comment (CrystallineCat) [reply website/server/controllers/api-v3/members.js]: The callee is throwing a normal `Error` instead of a HTTP `NotFound` to be independent of HTTP. Unfortunately, an unhandled `Error` will be turned into a generic 500 Internal Server Error and the message lost. This is fine for most API functions, but when the cause of the error can be an API parameter directly, I think the error message needs to be exposed to the user.

File: website/server/controllers/api-v3/members.js
Comment (crookedneighbor) [inline website/server/controllers/api-v3/members.js]: We should enumerate in the API docs all the interactions that can be queried.

**Comment by CrystallineCat** [comment]
@Alys @crookedneighbor @khipkin: Is there anything else I need to do to get this merged? I think I addressed all the comments. Apologies if I missed anything.

File: website/server/controllers/api-v3/members.js
Comment (crookedneighbor) [inline website/server/controllers/api-v3/members.js:288]: I'm not thrilled about the name of this url (the `objections-to` part), but I can't think of a better name :(

Comment (crookedneighbor) [reply website/server/controllers/api-v3/members.js:288]: Can we mark this route as beta so we can change it without worrying about breaking 3rd party integrations?

File: website/server/controllers/api-v3/members.js
Comment (crookedneighbor) [inline website/server/controllers/api-v3/members.js]: Can we fill out this documentation a bit? See https://github.com/HabitRPG/habitrpg/pull/7879/files#diff-cf049c3d45b117fee868e17845f1f130R8 for an example.

File: test/api/v3/integration/members/GET-objections_to_interaction.test.js
Comment (crookedneighbor) [inline test/api/v3/integration/members/GET-objections_to_interaction.test.js]: This should be a translation string with interpolations:

``` json
"uknownInteraction": "Uknown kind of interaction: \"<%= unknownInteraction %>\". Expected one of <%= interactions %>"
```

``` js
let unknownInteraction = req.params.interaction;
let interactions = 'send-private-message, transfer-gems';
t('uknownInteraction', { unknownInteraction, interactions });
```

File: test/api/v3/integration/members/GET-objections_to_interaction.test.js
Comment (crookedneighbor) [inline test/api/v3/integration/members/GET-objections_to_interaction.test.js]: Technically it returns an empty array, not nothing.

File: website/server/controllers/api-v3/members.js
Comment (crookedneighbor) [inline website/server/controllers/api-v3/members.js]: I feel weird about this. Feels like the check for valid params should happen much earlier, before we look up the receiver.

File: website/server/models/user/methods.js
Comment (crookedneighbor) [inline website/server/models/user/methods.js]: The `checks` and `knownInteractions` variables can be pull out of this function since they are constant and do not need to be recreated for every time this method is called. 

Maybe:

``` js
const INTERACTION_CHECKS = Object.freeze({ /* functions here */ });
const KNOWN_INTERACTIONS = Object.freeze(Object.keys(INTERACTION_CHECKS).filter(key => key !== 'always'));

This means you can also export `KNOWN_INTERACTIONS` for a params check in the route instead of throwing an error here.
```

File: test/api/v3/integration/members/GET-objections_to_interaction.test.js
Comment (crookedneighbor) [reply test/api/v3/integration/members/GET-objections_to_interaction.test.js]: If using my thought to export the known interactions:

``` js
t('uknownInteraction', { unknownInteraction, interactions: KNOWN_INTERACTIONS.join(',') });
```

File: website/server/models/user/methods.js
Comment (crookedneighbor) [inline website/server/models/user/methods.js]: Might be better to use lodash chaining for the following lines so we're not creating a ton of extra arrays

File: website/server/controllers/api-v3/members.js
Comment (crookedneighbor) [inline website/server/controllers/api-v3/members.js:284]: I think it should return an array of objection messages, rather than error messages.

**Review by crookedneighbor — CHANGES_REQUESTED** [review]

File: website/server/models/user/methods.js
Comment (crookedneighbor) [reply website/server/models/user/methods.js]: ^ How is this going?

File: website/server/controllers/api-v3/members.js
Comment (CrystallineCat) [reply website/server/controllers/api-v3/members.js:284]: I'm not sure what you mean by "objection messages".

File: website/server/controllers/api-v3/members.js
Comment (CrystallineCat) [reply website/server/controllers/api-v3/members.js:288]: I added "- BETA" to the comment; is anything else required to mark it as beta?

File: website/server/controllers/api-v3/members.js
Comment (CrystallineCat) [reply website/server/controllers/api-v3/members.js]: I'm not sure there is anything to add. The example is of a method with a much more complicated parameter structure; the current documentation for the new method already includes everything about its parameters. The only missing information is the exception, and I don't think committing to a specific exception here is a good idea.

File: test/api/v3/integration/members/GET-objections_to_interaction.test.js
Comment (CrystallineCat) [reply test/api/v3/integration/members/GET-objections_to_interaction.test.js]: I decided to move this to the parameter checking with req.checkParams. This provides a less useful error message, but is consistent with the other methods and I'm feeling uneasy about handling this check separately. Ideally, all the parameter checking would be done data-driven, but that's out of scope for this, I think.

File: website/server/models/user/methods.js
Comment (CrystallineCat) [reply website/server/models/user/methods.js]: I've converted it to lodash. I'm fairly new to JS and haven't used that library before, so I hope I did it the right way. :)

TBH, I'm not sure I'm that concerned about the arrays; the number is constant per call and I don't think the method is going to be a bottleneck, as it is likely only to be called in response to relatively infrequent UI events.

I realize I've gone against some of your requests, I've considered them over and over again and couldn't get myself to agree with them. Apologies if this causes any issues and thank you for your reviews. Also apologies for the huge delay... it took me being shocked from recent events to numb my fear of this conversation enough... :/ I'll gladly step back if you'd prefer someone else to work on this.

File: test/api/v3/integration/members/GET-objections_to_interaction.test.js
Comment (TheHollidayInn) [inline test/api/v3/integration/members/GET-objections_to_interaction.test.js:54]: You should added a test here to show that the route returns objections

File: website/server/controllers/api-v3/members.js
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/members.js:284]: I think his is saying that since this route returns a list of objects, you should change your wroding to say "Return an array of objections"

File: website/server/controllers/api-v3/members.js
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/members.js:288]: Why don't we just called the route `objections`? You are getting objections right?

File: website/server/models/user/methods.js
Comment (paglias) [inline website/server/models/user/methods.js:7]: Heads up, we're migrating to lodash v4 (https://github.com/HabitRPG/habitica/pull/8495) which includes many breaking changes (https://github.com/lodash/lodash/wiki/Changelog#v400). Before merging this PR should be checked and updated to make sure all lodash methods are compatible with v4.

**Review by paglias — CHANGES_REQUESTED** [review]

**Comment by SabreCat** [comment]
Continued in #8755!

**Pull request closed without merging**
